addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java

```diff
@@ -0,0 +1,693 @@
+import org.eclipse.smarthome.core.types.State;
+import org.openhab.binding.phc.PHCBindingConstants;
+import org.openhab.binding.phc.internal.PHCHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import gnu.io.CommPortIdentifier;
+import gnu.io.NoSuchPortException;
+import gnu.io.PortInUseException;
+import gnu.io.RXTXPort;
+import gnu.io.SerialPort;
+import gnu.io.SerialPortEvent;
+import gnu.io.SerialPortEventListener;
+import gnu.io.UnsupportedCommOperationException;
+
+/**
+ * The {@link PHCBridgeHandler} is responsible for handling the serial Communication to and from the PHC Modules.
+ *
+ * @author Jonas Hohaus - Initial contribution
+ */
+
+public class PHCBridgeHandler extends BaseBridgeHandler implements SerialPortEventListener {
+
+    private Logger logger = LoggerFactory.getLogger(PHCBridgeHandler.class);
+
+    private static final int BAUD = 19200;
+    private static final int SEND_RETRY_COUNT = 15; // max count to send the same message
+    private static final int SEND_RETRY_TIME = 80; // time to wait for an acknowledge before send the message again
+
+    private InputStream serialIn;
+    private OutputStream serialOut;
+    short lastReceivedCrc;
+    byte[] messageFragment;
+    RXTXPort serialPort;
+
+    private Map<String, Boolean> toggleMap = new HashMap<String, Boolean>();
+    private ConcurrentNavigableMap<Long, QueueObject> queue = new ConcurrentSkipListMap<Long, QueueObject>();
+
+    private byte emLedOutputState[] = new byte[32];
+    private byte amOutputState[] = new byte[32];
+
+    private List<Byte> modules = new ArrayList<Byte>();
+
+    public PHCBridgeHandler(Bridge phcBridge) {
+        super(phcBridge);
+    }
+
+    @Override
+    public void initialize() {
+        String port = ((String) getConfig().get(PHCBindingConstants.PORT));
+
+        try {
+            // find the given port
+            CommPortIdentifier portId = CommPortIdentifier.getPortIdentifier(port);
+
+            // initialize serial port
+            serialPort = portId.open(this.getClass().getName(), 2000); // owner, timeout
+            serialIn = serialPort.getInputStream();
+            // set port parameters
+            serialPort.setSerialPortParams(BAUD, SerialPort.DATABITS_8, SerialPort.STOPBITS_2, SerialPort.PARITY_NONE);
+            serialPort.setFlowControlMode(SerialPort.FLOWCONTROL_NONE);
+
+            serialPort.addEventListener(this);
+            // activate the DATA_AVAILABLE notifier
+            serialPort.notifyOnDataAvailable(true);
+            serialPort.setRTS(true);
+
+            // get the output stream
+            serialOut = serialPort.getOutputStream();
+
+            sendPorBroadcast((byte) 0xFF);
+
+            byte[] b = { 0x01 };
+            for (int j = 0; j <= 0x1F; j++) {
+                serialWrite(buildMessage((byte) j, 0, b, false));
+
+            }
+            updateStatus(ThingStatus.ONLINE);
+
+            scheduler.submit(new Runnable() {
```

Comment:
I just recently learned, but for these continuously running process we make the exception to create your own thread because otherwise you keep one of the precious threads of the pool in use for all of time.